### ISSUE TITLE
CPU hike when the RAIDZ1 group have two failed disk and Pool I/O in Suspended state

### ISSUE THREAD

**Issue by datacore-senthil** [openzfsonwindows/openzfs#98]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Windows Server 2022 Standard Evaluation
Distribution Version	|
Kernel Version	|
Architecture	|
OpenZFS Version	|

### Describe the problem you're observing
When the zpool state changes to I/O suspended state due to two disk failure/Unavailable in RAIDZ1 group , the CPU hikes to 100% and it gets reduced only when we bring back the failed disk and run zpool clear.
### Describe how to reproduce the problem
1.Create Zpool with RAIDZ1 group , a mirror slog with primarycache=all , sync=always
2. Run some IOs on the Zvols
3. while IO was going on , remove two disks from the RAIDZ1 group
4. The pool become suspended and the CPU will start to shootup
5. CPU will reduce only after adding the disks and run zpool clear

### Include any warning/errors/backtraces from the system logs
We run windows Xperf tool and noticed the CPU hike in cv_wait_io. Attached the Xperf analysis for reference
![image](https://user-images.githubusercontent.com/79847390/172568165-599a25b8-9869-40ce-b5aa-4e60a24a9960.png)

**Comment by lundman** [comment]
Hmm `spl_cv_wait()` doesn't have any loops so it's presumably dropping through it immediately, so we end up looping around
```
while (tx->tx_synced_txg < txg) {
		dprintf("broadcasting sync more "
		    "tx_synced=%llu waiting=%llu dp=%p\n",
		    tx->tx_synced_txg, tx->tx_sync_txg_waiting, dp);
		cv_broadcast(&tx->tx_sync_more_cv);
		cv_wait(&tx->tx_sync_done_cv, &tx->tx_sync_lock);
	}
```
Or something similar. I don't think I have a `txg_wait_synced_impl()` in my old ZFSin, but presumably it is similar to pasted above. Do we need to reset the event maybe so two consecutive calls to cv_wait wont just pass-through on the one trigger?

**Comment by datacore-senthil** [comment]
When we tried the same case in linux we didn't face this CPU hike. It looks like linux implementation of `cv_wait_io` which is called in `txg_wait_synced_impl()` is different from windows.

`static void
cv_wait_common(kcondvar_t *cvp, kmutex_t *mp, int state, int io)
{
	DEFINE_WAIT(wait);
	kmutex_t *m;

	ASSERT(cvp);
	ASSERT(mp);
	ASSERT(cvp->cv_magic == CV_MAGIC);
	ASSERT(mutex_owned(mp));
	atomic_inc(&cvp->cv_refs);

	m = READ_ONCE(cvp->cv_mutex);
	if (!m)
		m = xchg(&cvp->cv_mutex, mp);
	/* Ensure the same mutex is used by all callers */
	ASSERT(m == NULL || m == mp);

	prepare_to_wait_exclusive(&cvp->cv_event, &wait, state);
	atomic_inc(&cvp->cv_waiters);

	/*
	 * Mutex should be dropped after prepare_to_wait() this
	 * ensures we're linked in to the waiters list and avoids the
	 * race where 'cvp->cv_waiters > 0' but the list is empty.
	 */
	mutex_exit(mp);
	if (io)
		io_schedule();
	else
		schedule();`

In above function for experiment when we commented out the `io_schedule()` we able to see the CPU hike in linux as well

**Comment by lundman** [comment]
Oh ah I see, ok, on linux, if you want to manual wait, it looks like:
```
 for (;;) {
        add_wait_queue(&queue, &wait);
        set_current_state(TASK_INTERRUPTIBLE);
	if (condition)
	    break;
        schedule();
	remove_wait_queue(&queue, &wait);
	if (signal_pending(current))
	    return -ERESTARTSYS;
    }
    set_current_state(TASK_RUNNING);
```

So the `schedule()` call is in fact what sleeps. So you would definitely see CPU spinning if you take that call out.

**Comment by datacore-senthil** [comment]
@lundman  Can we try adding  below code in `cv_wait_io` to relinquish the CPU
`LARGE_INTEGER delay = { 0 };`
`delay.QuadPart = 0;`
`KeDelayExecutionThread(KernelMode, FALSE, &delay);`

**Comment by lundman** [comment]
Yeah, that would be worth trying, was there a `kpreempt()` call? don't remember if I fill it in with real code.

I wonder if the cause could be related to:
```
	int last_waiter =
	    result == STATUS_WAIT_0 + CV_BROADCAST &&
	    cvp->cv_waiters_count == 1;

	if (last_waiter)
		KeClearEvent(&cvp->cv_kevent[CV_BROADCAST]);

	atomic_dec_32(&cvp->cv_waiters_count);
```

Since we only "clear" the Event in some situations, but the `txg_wait()` loop is fast enough to simply not-wait, pass right through and CPU-spin over and over on it.  Could also be interesting to "clear it if we've spun X times" or something, just to rule it out as not relevant.

**Comment by datacore-senthil** [comment]
@lundman We tried calling `KeDelayExecutionThread' function after the `cv_wait_io` and looks like there is no CPU hike with the minimal testing.

The code you where mentioning is in `spl_cv_timedwait'. We see CPU hike in `cv_wait_io` which in turn calls `spl_cv_wait` . In this function also we clear event only when the 'cv_waiters_count' is zero.

**Comment by lundman** [comment]
Yeah, sorry, they are all essentially the same flow, but yes the `spl_cv_wait()` variant. The broadcast is supposed to wait for all readers to wake, but clearing the Event, but it doesn't actually know one reader from another reader, or if the same one has woken multiple times. But if you clear the Event and CPU goes down, that suggests there is something to fix there
